### PR TITLE
babel-parser(ts): Throw recoverable for duplicates access modifier

### LINKED ISSUE #12774: babel-parser(ts): Unfriendly error for duplicates access modifier
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

Syntax error `Unexpected token` is thrown.

- [REPL](https://babeljs.io/repl#?browsers=safari%2012&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBiD29oG8BQ0MAcBOBLAbmAC4Cm0mArgEYi7DQBmiA3KgL6pA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.12.14&externalPlugins=)
**Input Code**

```ts
class Foo {
   private public foo;
}
```

**Expected behavior**

Syntax recoverable error `Accessibility modifier already seen.` is thrown. Like [TS](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ1oAcBOBLAbmAC4CmWArgEYi7DQBmiA3KgL6pA). Also only the first access modifier should be included in AST.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

**Environment**

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by sosukesuzuki**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12774
| Patch: Bug Fix?          | Y
| Tests Added + Pass?      | Yes
| License                  | MIT

### CODE REVIEW

**Review by kaicataldo — APPROVED** [review]

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -31,14 +31,13 @@ import type { ExpressionErrors } from "../../parser/util";
 import { PARAM } from "../../util/production-parameter";
 import { Errors } from "../../parser/error";
 
+type AccessibilityModifier = "public" | "private" | "protected";
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:34]: You can access `N.Accessibility` from https://github.com/babel/babel/blob/759a82abb1f2ea64c03dc9b298aa03950690fa27/packages/babel-parser/src/types.js#L661

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -2849,4 +2863,12 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         this.state.inAbstractClass = oldInAbstractClass;
       }
     }
+
+    tsIsAccessModifier(modifier: string): boolean {
+      return (
+        modifier === "private" ||
+        modifier === "public" ||
+        modifier === "protected"
+      );
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:2872]: Consider extract this function outside of the class, (it's static anyway) so we can use [predicate functions](https://flow.org/en/docs/types/functions/#toc-predicate-functions):
```flow
function tsIsAccessModifier(modifier: string): boolean %checks {
  return (
    modifier === "private" || modifier === "public" || modifier === "protected"
  );
}
```
In this way Flow will infer that `modifier` must be an `Accessibility` enum when `tsIsAccessModifier(modifier)` is `true`.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 3]
```diff
@@ -194,20 +194,32 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
 
         if (!modifier) break;
 
-        if (Object.hasOwnProperty.call(modified, modifier)) {
-          this.raise(startPos, TSErrors.DuplicateModifier, modifier);
+        if (this.tsIsAccessModifier(modifier)) {
+          if (modified.accessibility) {
+            this.raise(startPos, TSErrors.DuplicateAccessibilityModifier);
+          } else {
+            // $FlowIgnore
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:214]: `// $FlowIgnore` should be avoided, see explainers below.

**Review by JLHwung — APPROVED** [review]
LGTM with some typing nits.

File: packages/babel-parser/src/plugins/typescript/index.js (hunk 2 above)
Comment (sosukesuzuki) [reply packages/babel-parser/src/plugins/typescript/index.js:2872]: I didn't know how to define predicate function with Flow, Thanks!

**Review by existentialism — APPROVED** [review]

**Pull request merged**